File: app/src/main/java/wannabit/io/cosmostaion/fragment/SendStep0Fragment.java
package wannabit.io.cosmostaion.fragment;

import android.app.Activity;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.Nullable;

import com.google.zxing.integration.android.IntentIntegrator;
import com.google.zxing.integration.android.IntentResult;

import java.util.concurrent.TimeUnit;

import io.grpc.stub.StreamObserver;
import starnamed.x.starname.v1beta1.QueryGrpc;
import starnamed.x.starname.v1beta1.QueryOuterClass;
import wannabit.io.cosmostaion.R;
import wannabit.io.cosmostaion.activities.SendActivity;
import wannabit.io.cosmostaion.base.BaseChain;
import wannabit.io.cosmostaion.base.BaseFragment;
import wannabit.io.cosmostaion.dialog.Dialog_StarName_Confirm;
import wannabit.io.cosmostaion.network.ChannelBuilder;
import wannabit.io.cosmostaion.utils.WDp;
import wannabit.io.cosmostaion.utils.WKey;
import wannabit.io.cosmostaion.utils.WUtil;

import static wannabit.io.cosmostaion.base.BaseChain.AKASH_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.ALTHEA_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.AXELAR_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.BAND_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.BNB_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.BNB_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.CERTIK_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.CERTIK_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.COSMOS_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.COSMOS_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.CRYPTO_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.EMONEY_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.FETCHAI_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.IOV_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.IOV_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.IRIS_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.IRIS_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.JUNO_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.KAVA_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.KAVA_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.KI_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.MEDI_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.MEDI_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.OKEX_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.OK_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.OSMOSIS_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.PERSIS_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.RIZON_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.RIZON_TEST;
import static wannabit.io.cosmostaion.base.BaseChain.SECRET_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.SENTINEL_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.SIF_MAIN;
import static wannabit.io.cosmostaion.base.BaseChain.UMEE_TEST;
import static wannabit.io.cosmostaion.network.ChannelBuilder.TIME_OUT;

public class SendStep0Fragment extends BaseFragment implements View.OnClickListener {
    public final static int SELECT_STAR_NAME_ADDRESS = 9102;

    private EditText        mAddressInput;
    private Button          mCancel, mNextBtn;
    private LinearLayout    mStarNameLayer;
    private LinearLayout    mBtnQr, mBtnPaste, mBtnHistory;

    public static SendStep0Fragment newInstance(Bundle bundle) {
        SendStep0Fragment fragment = new SendStep0Fragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_send_step0, container, false);
        mAddressInput = rootView.findViewById(R.id.receiver_account);
        mNextBtn = rootView.findViewById(R.id.btn_next);
        mCancel = rootView.findViewById(R.id.btn_cancel);
        mStarNameLayer = rootView.findViewById(R.id.starname_layer);

        mBtnQr = rootView.findViewById(R.id.btn_qr);
        mBtnPaste = rootView.findViewById(R.id.btn_paste);
        mBtnHistory = rootView.findViewById(R.id.btn_history);
        mBtnHistory.setVisibility(View.GONE);

        mCancel.setOnClickListener(this);
        mNextBtn.setOnClickListener(this);
        mBtnQr.setOnClickListener(this);
        mBtnPaste.setOnClickListener(this);
        mBtnHistory.setOnClickListener(this);

        rootView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            private boolean alreadyOpen;
            private final int defaultKeyboardHeightDP = 100;
            private final int EstimatedKeyboardDP = defaultKeyboardHeightDP + (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP ? 48 : 0);
            private final Rect rect = new Rect();

            @Override
            public void onGlobalLayout() {
                int estimatedKeyboardHeight = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, EstimatedKeyboardDP, rootView.getResources().getDisplayMetrics());
                rootView.getWindowVisibleDisplayFrame(rect);
                int heightDiff = rootView.getRootView().getHeight() - (rect.bottom - rect.top);
                boolean isShown = heightDiff >= estimatedKeyboardHeight;
                if (isShown == alreadyOpen) {
                    return;
                }
                alreadyOpen = isShown;
                if (alreadyOpen) {
                    mStarNameLayer.setVisibility(View.GONE);
                } else {
                    new Handler().postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            mStarNameLayer.setVisibility(View.VISIBLE);
                        }
                    },100);
                }
            }
        });
        mStarNameLayer.setVisibility(View.VISIBLE);
        return rootView;
    }

    @Override
    public void onClick(View v) {
        if (v.equals(mNextBtn)) {
            String userInput = mAddressInput.getText().toString().trim();

            if (getSActivity().mAccount.address.equals(userInput)) {
                Toast.makeText(getContext(), R.string.error_self_sending, Toast.LENGTH_SHORT).show();
                return;
            }

            if (WUtil.isValidStarName(userInput.toLowerCase())) {
                onCheckNameService(userInput.toLowerCase(), getSActivity().mBaseChain);
                return;
            }

            if (WDp.isValidChainAddress(getSActivity().mBaseChain, userInput)) {
                getSActivity().mToAddress = userInput;
                getSActivity().onNextStep();
            } else {
                Toast.makeText(getContext(), R.string.error_invalid_address_target, Toast.LENGTH_SHORT).show();
            }

        } else if (v.equals(mCancel)) {
            getSActivity().onBeforeStep();

        } else if (v.equals(mBtnQr)) {
            IntentIntegrator integrator = IntentIntegrator.forSupportFragment(this);
            integrator.setOrientationLocked(true);
            integrator.initiateScan();

        } else if (v.equals(mBtnPaste)) {
            ClipboardManager clipboard = (ClipboardManager)getSActivity().getSystemService(Context.CLIPBOARD_SERVICE);
            if(clipboard.getPrimaryClip() != null && clipboard.getPrimaryClip().getItemCount() > 0) {
                String userPaste = clipboard.getPrimaryClip().getItemAt(0).coerceToText(getSActivity()).toString().trim();
                if(TextUtils.isEmpty(userPaste)) {
                    Toast.makeText(getSActivity(), R.string.error_clipboard_no_data, Toast.LENGTH_SHORT).show();
                    return;
                }
                mAddressInput.setText(userPaste);
                mAddressInput.setSelection(mAddressInput.getText().length());

            } else {
                Toast.makeText(getSActivity(), R.string.error_clipboard_no_data, Toast.LENGTH_SHORT).show();
            }


        } else if (v.equals(mBtnHistory)) {
            Toast.makeText(getSActivity(), R.string.error_prepare, Toast.LENGTH_SHORT).show();

        }
    }

    private SendActivity getSActivity() {
        return (SendActivity)getBaseActivity();
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == SELECT_STAR_NAME_ADDRESS) {
            if (resultCode == Activity.RESULT_OK) {
                getSActivity().mStarName = data.getStringExtra("starname");
                getSActivity().mToAddress = data.getStringExtra("originAddress");
                getSActivity().onNextStep();
            }

        } else {
            IntentResult result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);
            if(result != null) {
                if(result.getContents() != null) {
                    mAddressInput.setText(result.getContents().trim());
                    mAddressInput.setSelection(mAddressInput.getText().length());
                }
            } else {
                super.onActivityResult(requestCode, resultCode, data);
            }
        }
    }

    private void onCheckNameService(String userInput, BaseChain chain) {
        QueryGrpc.QueryStub mStub = QueryGrpc.newStub(ChannelBuilder.getChain(IOV_MAIN)).withDeadlineAfter(TIME_OUT, TimeUnit.SECONDS);
        QueryOuterClass.QueryStarnameRequest request = QueryOuterClass.QueryStarnameRequest.newBuilder().setStarname(userInput).build();
        mStub.starname(request, new StreamObserver<QueryOuterClass.QueryStarnameResponse>() {
            @Override
            public void onNext(QueryOuterClass.QueryStarnameResponse value) {
                new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        final String matchAddress = WUtil.checkStarnameWithResource(chain, value.getAccount().getResourcesList());
                        if (TextUtils.isEmpty(matchAddress)) {
                            Toast.makeText(getContext(), R.string.error_no_mattched_starname, Toast.LENGTH_SHORT).show();
                            return;
                        }

                        if (getSActivity().mAccount.address.equals(matchAddress)) {
                            Toast.makeText(getContext(), R.string.error_no_mattched_starname, Toast.LENGTH_SHORT).show();
                            return;
                        }

                        Bundle bundle = new Bundle();
                        bundle.putString("starname", userInput);
                        bundle.putString("originAddress", matchAddress);
                        Dialog_StarName_Confirm dialog = Dialog_StarName_Confirm.newInstance(bundle);
                        dialog.setCancelable(true);
                        dialog.setTargetFragment(SendStep0Fragment.this, SELECT_STAR_NAME_ADDRESS);
                        getFragmentManager().beginTransaction().add(dialog, "dialog").commitNowAllowingStateLoss();
                    }
                }, 0);

            }

            @Override
            public void onError(Throwable t) {
                new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(getContext(), R.string.error_invalide_starname, Toast.LENGTH_SHORT).show();
                    }
                }, 0);
            }

            @Override
            public void onCompleted() { }
        });


    }
}

File: app/src/main/java/wannabit/io/cosmostaion/fragment/txs/common/SendStep0Fragment.kt
package wannabit.io.cosmostaion.fragment.txs.common

import android.app.Activity
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Rect
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.text.Html
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.fragment.app.viewModels
import com.google.zxing.client.android.Intents
import com.google.zxing.integration.android.IntentIntegrator
import wannabit.io.cosmostaion.R
import wannabit.io.cosmostaion.activities.QRcodeActivity
import wannabit.io.cosmostaion.activities.txs.common.SendActivity
import wannabit.io.cosmostaion.base.BaseChain
import wannabit.io.cosmostaion.base.BaseConstant
import wannabit.io.cosmostaion.base.BaseFragment
import wannabit.io.cosmostaion.base.chains.ChainConfig
import wannabit.io.cosmostaion.base.chains.ChainFactory
import wannabit.io.cosmostaion.dao.Account
import wannabit.io.cosmostaion.dao.Asset
import wannabit.io.cosmostaion.dao.MintscanToken
import wannabit.io.cosmostaion.dao.NameService
import wannabit.io.cosmostaion.dao.NameService.NameServiceType
import wannabit.io.cosmostaion.databinding.FragmentSendStep0Binding
import wannabit.io.cosmostaion.dialog.CommonAlertDialog
import wannabit.io.cosmostaion.dialog.IBCReceiveAccountsDialog
import wannabit.io.cosmostaion.dialog.NameConfirmDialog
import wannabit.io.cosmostaion.dialog.SelectChainListDialog
import wannabit.io.cosmostaion.model.SendViewModel
import wannabit.io.cosmostaion.utils.WDp
import wannabit.io.cosmostaion.utils.WUtil
import wannabit.io.cosmostaion.utils.makeToast
import wannabit.io.cosmostaion.utils.visibleOrGone

class SendStep0Fragment : BaseFragment() {

    private var _binding: FragmentSendStep0Binding? = null
    private val binding get() = _binding!!
    private val sendViewModel: SendViewModel by viewModels()

    private val mToSendableChains = ArrayList<ChainConfig>()
    private var mToSendChainConfig: ChainConfig? = null
    private var mToAccountList: ArrayList<Account>? = null
    private var mAsset: Asset? = null
    private var mMintscanToken: MintscanToken? = null

    private val mNameServices = ArrayList<NameService>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentSendStep0Binding.inflate(layoutInflater, container, false)

        mAsset = baseDao.getAsset(getSActivity().mChainConfig, getSActivity().mDenom)
        mMintscanToken = baseDao.getCw20Asset(getSActivity().mChainConfig, getSActivity().mDenom)
        mToSendableChains.add(getSActivity().mChainConfig)

        val allChainConfig = ChainFactory.SUPPRT_CONFIG()
        for (asset in baseDao.mAssets) {
            if (mAsset != null) {
                if (asset.chain.equals(getSActivity().mChainConfig.chainName(), ignoreCase = true) && asset.denom.equals(getSActivity().mDenom, ignoreCase = true)) {
                    for (chainConfig in allChainConfig) {
                        if (chainConfig.chainName().equals(asset.beforeChain(getSActivity().mChainConfig), ignoreCase = true) && !mToSendableChains.contains(chainConfig)) {
                            mToSendableChains.add(chainConfig)
                        }
                    }
                } else if (asset.counter_party != null && asset.counter_party.denom.equals(getSActivity().mDenom, ignoreCase = true)) {
                    for (chainConfig in allChainConfig) {
                        if (chainConfig.chainName().equals(asset.chain, ignoreCase = true) && !mToSendableChains.contains(chainConfig)) {
                            mToSendableChains.add(chainConfig)
                        }
                    }
                }

            } else if (mMintscanToken != null) {
                if (asset.counter_party != null && asset.counter_party.denom.equals(mMintscanToken!!.address, ignoreCase = true)) {
                    for (chainConfig in allChainConfig) {
                        if (chainConfig.chainName().equals(asset.chain, ignoreCase = true)) {
                            mToSendableChains.add(chainConfig)
                        }
                    }
                }
            }
        }
        onSortToChain()
        onLoadObserve()

        mToSendChainConfig = mToSendableChains[0]
        mToSendChainConfig?.let {
            mToAccountList = baseDao.onSelectAccountsExceptSelfByChain(it.baseChain(), getSActivity().mAccount)
        }
        return binding.root
    }

    private fun onSetIbcImgLayout() {
        binding.root.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            private var alreadyOpen: Boolean = false
            private val defaultKeyboardHeightDP: Int = 100
            private val EstimatedKeyboardDP = defaultKeyboardHeightDP + if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) 48 else 0
            private val rect = Rect()
            override fun onGlobalLayout() {
                val estimatedKeyboardHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, EstimatedKeyboardDP.toFloat(), binding.root.resources.displayMetrics).toInt()
                binding.root.getWindowVisibleDisplayFrame(rect)
                val heightDiff: Int = binding.root.rootView.height - (rect.bottom - rect.top)
                val isShown = heightDiff >= estimatedKeyboardHeight
                if (isShown == alreadyOpen) return

                alreadyOpen = isShown
                if (alreadyOpen) {
                    binding.ibcLayer.visibility = View.GONE
                } else {
                    Handler().postDelayed({
                        mToSendChainConfig?.let {
                            binding.ibcLayer.visibleOrGone(!it.baseChain().equals(getSActivity().mBaseChain))
                        }
                    }, 100)
                }
            }
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        onUpdateChainView()
        onSetIbcImgLayout()
        onClick()
    }

    private fun onUpdateChainView() {
        mToSendChainConfig?.let { chainConfig ->
            binding.apply {
                imgToChain.setImageResource(chainConfig.chainImg())
                txtToChain.text = chainConfig.chainTitleToUp()
                txtToChain.setTextColor(ContextCompat.getColor(requireContext(), chainConfig.chainColor()))
                receiverAccount.setText("")
                ibcLayer.visibleOrGone(!chainConfig.baseChain().equals(getSActivity().mBaseChain))

//                if (!chainConfig.baseChain().equals(getSActivity().mBaseChain)) {
//                    val assetPath = WDp.getAssetPath(baseDao, getSActivity().mChainConfig, mToSendChainConfig, getSActivity().mDenom)
//                    sendViewModel.loadPendingTotal(getSActivity().mChainConfig, assetPath)
//                    sendViewModel.loadLastPing(requireContext(), assetPath, baseDao.chainIdGrpc)
//                }
            }
        }
    }

    private fun onClick() {
        binding.apply {
            btnToChainList.setOnClickListener {
                Bundle().apply {
                    putSerializable(SelectChainListDialog.TO_SENDABLE_CHAIN_CONFIG_BUNDLE_KEY, mToSendableChains)
                    putInt(SelectChainListDialog.SELECT_CHAIN_LIST_BUNDLE_KEY, SelectChainListDialog.SELECT_IBC_CHAIN_VALUE)
                    SelectChainListDialog.newInstance(this).show(parentFragmentManager, SelectChainListDialog::class.java.name)

                    parentFragmentManager.setFragmentResultListener(SelectChainListDialog.SELECT_CHAIN_LIST_BUNDLE_KEY, requireActivity()) { _, result ->
                        mToSendChainConfig = mToSendableChains[result.getInt(BaseConstant.POSITION)]
                        mToSendChainConfig?.let {
                            mToAccountList = baseDao.onSelectAccountsExceptSelfByChain(it.baseChain(), getSActivity().mAccount)
                            onUpdateChainView()
                        }
                    }
                }
                getSActivity().onHideKeyboard()
            }

            btnWallet.setOnClickListener {
                if (mToSendChainConfig == null || (mToAccountList?.size?: 0) <= 0) {
                    requireContext().makeToast(R.string.error_no_wallet_this_chain)
                    return@setOnClickListener
                } else {
                    Bundle().apply {
                        putSerializable(IBCReceiveAccountsDialog.ACCOUNTS_BUNDLE_KEY, mToAccountList)
                        IBCReceiveAccountsDialog.newInstance(this).show(parentFragmentManager, IBCReceiveAccountsDialog::class.java.name)

                        parentFragmentManager.setFragmentResultListener(IBCReceiveAccountsDialog.IBC_RECEIVE_ACCOUNTS_BUNDLE_KEY, requireActivity()) { _, result ->
                            mToAccountList?.let {
                                binding.receiverAccount.setText(it[result.getInt(BaseConstant.POSITION)].address)
                            }
                        }
                    }
                }
            }

            btnQr.setOnClickListener {
                val integrator = IntentIntegrator.forSupportFragment(this@SendStep0Fragment)
                integrator.setOrientationLocked(true)
                integrator.captureActivity = QRcodeActivity::class.java
                qrCodeResultLauncher.launch(integrator.createScanIntent())
            }

            btnPaste.setOnClickListener {
                val clipboard = getSActivity().getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                if (clipboard.primaryClip != null && clipboard.primaryClip!!.itemCount > 0) {
                    val userPaste = clipboard.primaryClip!!.getItemAt(0).coerceToText(requireContext()).toString().trim()
                    if (userPaste.isEmpty()) {
                        requireContext().makeToast(R.string.error_clipboard_no_data)
                        return@setOnClickListener
                    } else {
                        receiverAccount.setText(userPaste)
                        receiverAccount.setSelection(receiverAccount.text.length)
                    }
                } else {
                    requireContext().makeToast(R.string.error_clipboard_no_data)
                    return@setOnClickListener
                }
            }

            btnCancel.setOnClickListener {
                getSActivity().onBeforeStep()
            }

            btnNext.setOnClickListener {
                mNameServices.clear()
                receiverAccount.text.trim().toString().apply {
                    if (getSActivity().mAccount.address.equals(this, ignoreCase = true)) {
                        requireContext().makeToast(R.string.error_self_sending)
                        return@setOnClickListener
                    }

                    if (WUtil.isValidStarName(this.lowercase())) {
                        mToSendChainConfig?.let {
                            onPathSetting()
                            sendViewModel.loadResources(it, this.lowercase())
                            return@setOnClickListener
                        }
                    }

                    if (WDp.isValidChainAddress(mToSendChainConfig, this)) {
                        if (!isExchangeAddress(this)) {
                            CommonAlertDialog.showSingleButton(
                                activity,
                                Html.fromHtml("<font color=\"#f31963\">" + getString(R.string.str_empty_warnning_title) + "</font>", Html.FROM_HTML_MODE_COMPACT),
                                getString(R.string.error_exchange_address_msg),
                                getString(R.string.str_confirm),
                                null,
                                false
                            )
                            return@setOnClickListener
                        }
                        onPathSetting()
                        getSActivity().mToAddress = this
                        getSActivity().onNextStep()
                    } else {
                        mToSendChainConfig?.let {
                            sendViewModel.loadIcnsAddress(it, userInput(it))
                        }
                        onPathSetting()
                    }
                }
            }
        }
    }

    private val qrCodeResultLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            binding.receiverAccount.setText(result.data?.getStringExtra(Intents.Scan.RESULT)?.trim())
            binding.receiverAccount.setSelection(binding.receiverAccount.text.length)
        }
    }

    private fun onLoadObserve() {
        sendViewModel.resource.observe(requireActivity()) { resources ->
            mToSendChainConfig?.let {
                val matchAddress = resources[0].address
                if (matchAddress.isEmpty()) {
                    requireContext().makeToast(R.string.error_no_mattched_starname)
                    return@observe
                }
                onSetICNS(it, resources, matchAddress)
            }
        }

        sendViewModel.nameServices.observe(requireActivity()) { nameServices ->
            mToSendChainConfig?.let {
                if (nameServices.size <= 0) {
                    requireContext().makeToast(R.string.error_invalid_icns)
                    return@observe
                }
                if (nameServices.size == 2 && nameServices[0].address.equals(nameServices[1].address, ignoreCase = true)) {
                    nameServices[0].type = NameServiceType.ICNS_STARGAZE
                    nameServices.removeAt(nameServices.size - 1)
                }
                onSetICNS(it, nameServices, nameServices[0].address)
            }
        }

//        sendViewModel.lastPing.observe(requireActivity()) {
//            binding.lastPingUpdate.text = WDp.getTimeGap(requireContext(), it)
//        }
//
//        sendViewModel.pendingTotal.observe(requireActivity()) {
//            binding.pendingCount.text = it.toString()
//        }
    }

    private fun onPathSetting() {
        mToSendChainConfig?.let { chainConfig ->
            if (getSActivity().mBaseChain == chainConfig.baseChain()) {
                if (mMintscanToken == null) getSActivity().mTxType = BaseConstant.CONST_PW_TX_SIMPLE_SEND
                else {
                    mMintscanToken?.let {
                        if (it.address.startsWith("0x")) getSActivity().mTxType = BaseConstant.CONST_PW_TX_EVM_TRANSFER
                        else getSActivity().mTxType = BaseConstant.CONST_PW_TX_EXECUTE_CONTRACT
                    }
                }

            } else {
                if (mAsset != null) getSActivity().mTxType = BaseConstant.CONST_PW_TX_IBC_TRANSFER
                else if (mMintscanToken != null) getSActivity().mTxType = BaseConstant.CONST_PW_TX_IBC_CONTRACT

                getSActivity().mAssetPath = WDp.getAssetPath(baseDao, getSActivity().mChainConfig, mToSendChainConfig, getSActivity().mDenom)
            }
        }
        getSActivity().mAsset = mAsset
        getSActivity().mMintscanToken = mMintscanToken
    }

    private fun isExchangeAddress(userInput: String): Boolean {
        return !(WUtil.getExchangeAddressList().contains(userInput) && !userInput.startsWith(getSActivity().mChainConfig.addressPrefix()))
    }

    private fun userInput(chainConfig: ChainConfig): String {
        chainConfig.let {
            binding.receiverAccount.text.trim().toString().let { userInput ->
                if (userInput.contains("." + it.addressPrefix())) {
                    return userInput
                } else if (userInput.contains(".")) {
                    return userInput + it.addressPrefix()
                } else {
                    return userInput + "." + it.addressPrefix()
                }
            }
        }
    }

    private fun onSetICNS(chainConfig: ChainConfig, nameServiceList: ArrayList<NameService>, matchAddress: String) {
        Bundle().apply {
            putSerializable(NameConfirmDialog.MATCH_NAME_SERVICE_BUNDLE_KEY, nameServiceList)
            val dialog = NameConfirmDialog.newInstance(this)
            dialog.show(parentFragmentManager, NameConfirmDialog::class.java.name)

            parentFragmentManager.setFragmentResultListener(NameConfirmDialog.CONFIRM_BUNDLE_KEY, requireActivity()) { _, result ->
                if (getSActivity().mAccount.address.equals(matchAddress, ignoreCase = true)) {
                    requireContext().makeToast(R.string.error_self_sending)
                    dialog.dismiss()
                    return@setFragmentResultListener
                }

                if (!WDp.isValidChainAddress(chainConfig, matchAddress)) {
                    requireContext().makeToast(R.string.error_invalid_icns_address)
                    dialog.dismiss()
                    return@setFragmentResultListener
                }

                getSActivity().mToAddress = nameServiceList[result.getInt(BaseConstant.POSITION)].address
                getSActivity().onNextStep()
            }
        }
    }

    private fun onSortToChain() {
        mToSendableChains.sortWith(Comparator { o1, o2 ->
            if (o1.baseChain() == getSActivity().mBaseChain) return@Comparator -1
            if (o2.baseChain() == getSActivity().mBaseChain) return@Comparator 1
            if (o1.baseChain() == BaseChain.COSMOS_MAIN) return@Comparator -1
            if (o2.baseChain() == BaseChain.COSMOS_MAIN) return@Comparator 1
            if (o1.baseChain() == BaseChain.OSMOSIS_MAIN) return@Comparator -1
            if (o2.baseChain() == BaseChain.OSMOSIS_MAIN) return@Comparator 1
            0
        })
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }

    private fun getSActivity(): SendActivity {
        return baseActivity as SendActivity
    }
}